File: sdk-dotnet/src/HealthFlow.Hfcx.Sdk/Recipient/EgyptianBundleValidator.cs
// Copyright (c) HealthFlow Medical HCX. Licensed under Apache 2.0.

using System;
using System.Text.Json;
using HealthFlow.Hfcx.Sdk.Exceptions;
using HealthFlow.Hfcx.Sdk.Validators;

namespace HealthFlow.Hfcx.Sdk.Recipient;

/// <summary>
/// Walks a FHIR Bundle and runs the Egyptian field-level validators
/// (National ID, phone, IBAN). Sister to Java's
/// <c>EgyptianBundleValidator</c> and Python's
/// <c>EgyptianBundleValidator</c>.
/// </summary>
public sealed class EgyptianBundleValidator
{
    /// <summary>Validate the Bundle JSON; throws on first violation.</summary>
    public void Validate(string? bundleJson)
    {
        if (string.IsNullOrWhiteSpace(bundleJson))
        {
            return;
        }

        JsonDocument doc;
        try
        {
            doc = JsonDocument.Parse(bundleJson);
        }
        catch (JsonException)
        {
            // FhirValidator runs first; fail-secure no-op here.
            return;
        }

        using (doc)
        {
            var root = doc.RootElement;
            if (root.ValueKind != JsonValueKind.Object) return;
            if (!root.TryGetProperty("entry", out var entries)
                || entries.ValueKind != JsonValueKind.Array)
            {
                return;
            }

            foreach (var wrapper in entries.EnumerateArray())
            {
                if (wrapper.ValueKind != JsonValueKind.Object) continue;
                if (!wrapper.TryGetProperty("resource", out var resource)
                    || resource.ValueKind != JsonValueKind.Object)
                {
                    continue;
                }

                var rt = ReadString(resource, "resourceType");
                if (string.Equals(rt, "Patient", StringComparison.Ordinal))
                {
                    ValidatePatient(resource);
                }
                else if (string.Equals(rt, "Organization", StringComparison.Ordinal))
                {
                    ValidateOrganization(resource);
                }
            }
        }
    }

    private static void ValidatePatient(JsonElement patient)
    {
        if (patient.TryGetProperty("identifier", out var identifiers)
            && identifiers.ValueKind == JsonValueKind.Array)
        {
            foreach (var ident in identifiers.EnumerateArray())
            {
                if (ident.ValueKind != JsonValueKind.Object) continue;
                if (string.Equals(ReadString(ident, "system"), FhirValidator.NationalIdSystem, StringComparison.Ordinal))
                {
                    var value = ReadString(ident, "value");
                    if (!EgyptianNationalIdValidator.IsValid(value))
                    {
                        throw new NationalIdInvalidException(
                            $"Patient National-ID identifier value '{value}' is not a valid Egyptian National ID");
                    }
                }
            }
        }

        if (patient.TryGetProperty("telecom", out var telecom)
            && telecom.ValueKind == JsonValueKind.Array)
        {
            foreach (var contact in telecom.EnumerateArray())
            {
                if (contact.ValueKind != JsonValueKind.Object) continue;
                if (string.Equals(ReadString(contact, "system"), "phone", StringComparison.Ordinal))
                {
                    var value = ReadString(contact, "value");
                    if (!EgyptianPhoneValidator.IsValid(value))
                    {
                        throw new PhoneInvalidException(
                            $"Patient phone '{value}' is not a valid Egyptian mobile number");
                    }
                }
            }
        }
    }

    private static void ValidateOrganization(JsonElement org)
    {
        if (!org.TryGetProperty("identifier", out var identifiers)
            || identifiers.ValueKind != JsonValueKind.Array)
        {
            return;
        }

        foreach (var ident in identifiers.EnumerateArray())
        {
            if (ident.ValueKind != JsonValueKind.Object) continue;
            var system = ReadString(ident, "system");
            if (system is not null
                && system.Contains("iban", StringComparison.Ordinal))
            {
                var value = ReadString(ident, "value");
                if (!EgyptianIbanValidator.IsValid(value))
                {
                    throw new IbanInvalidException(
                        $"Organization IBAN '{value}' is not a valid Egyptian IBAN");
                }
            }
        }
    }

    private static string? ReadString(JsonElement element, string field)
    {
        if (element.ValueKind != JsonValueKind.Object) return null;
        if (!element.TryGetProperty(field, out var value)) return null;
        return value.ValueKind == JsonValueKind.String ? value.GetString() : null;
    }
}

File: sdk-javascript/src/recipient/EgyptianBundleValidator.ts
// Copyright (c) HealthFlow Medical HCX. Licensed under Apache 2.0.

import {
  IbanInvalidError,
  NationalIdInvalidError,
  PhoneInvalidError,
} from '../exceptions/HfcxError.js';
import { isValidEgyptianIban } from '../validators/egyptianIban.js';
import { isValidEgyptianNationalId } from '../validators/egyptianNationalId.js';
import { isValidEgyptianPhone } from '../validators/egyptianPhone.js';
import { NATIONAL_ID_SYSTEM } from './FhirValidator.js';

/**
 * Walks a FHIR Bundle and runs the Egyptian field-level validators
 * (National ID, phone, IBAN). Sister to Java's
 * `EgyptianBundleValidator`, Python's `EgyptianBundleValidator`, and
 * .NET's `EgyptianBundleValidator`.
 */
export class EgyptianBundleValidator {
  validate(bundleJson: string | null | undefined): void {
    if (!bundleJson || !bundleJson.trim()) return;

    let root: unknown;
    try {
      root = JSON.parse(bundleJson);
    } catch {
      // FhirValidator runs first; fail-secure no-op here.
      return;
    }
    if (!root || typeof root !== 'object' || Array.isArray(root)) return;
    const entries = (root as { entry?: unknown }).entry;
    if (!Array.isArray(entries)) return;

    for (const wrapper of entries) {
      if (!wrapper || typeof wrapper !== 'object') continue;
      const resource = (wrapper as { resource?: unknown }).resource;
      if (!resource || typeof resource !== 'object') continue;
      const rt = (resource as { resourceType?: unknown }).resourceType;
      if (rt === 'Patient') {
        validatePatient(resource as Record<string, unknown>);
      } else if (rt === 'Organization') {
        validateOrganisation(resource as Record<string, unknown>);
      }
    }
  }
}

function validatePatient(patient: Record<string, unknown>): void {
  const identifiers = patient.identifier;
  if (Array.isArray(identifiers)) {
    for (const ident of identifiers) {
      if (
        !ident ||
        typeof ident !== 'object' ||
        (ident as { system?: unknown }).system !== NATIONAL_ID_SYSTEM
      ) {
        continue;
      }
      const value = (ident as { value?: unknown }).value;
      const v = typeof value === 'string' ? value : undefined;
      if (!isValidEgyptianNationalId(v)) {
        throw new NationalIdInvalidError(
          `Patient National-ID identifier value '${v}' is not a valid Egyptian National ID`,
        );
      }
    }
  }
  const telecom = patient.telecom;
  if (Array.isArray(telecom)) {
    for (const contact of telecom) {
      if (
        !contact ||
        typeof contact !== 'object' ||
        (contact as { system?: unknown }).system !== 'phone'
      ) {
        continue;
      }
      const value = (contact as { value?: unknown }).value;
      const v = typeof value === 'string' ? value : undefined;
      if (!isValidEgyptianPhone(v)) {
        throw new PhoneInvalidError(`Patient phone '${v}' is not a valid Egyptian mobile number`);
      }
    }
  }
}

function validateOrganisation(org: Record<string, unknown>): void {
  const identifiers = org.identifier;
  if (!Array.isArray(identifiers)) return;
  for (const ident of identifiers) {
    if (!ident || typeof ident !== 'object') continue;
    const system = (ident as { system?: unknown }).system;
    if (typeof system === 'string' && system.includes('iban')) {
      const value = (ident as { value?: unknown }).value;
      const v = typeof value === 'string' ? value : undefined;
      if (!isValidEgyptianIban(v)) {
        throw new IbanInvalidError(`Organization IBAN '${v}' is not a valid Egyptian IBAN`);
      }
    }
  }
}
